File: templates/objc/cudnn_wrapper.m
#import "cudnn_wrapper.h"

@implementation CUDNNWrapper {
    id<MTLDevice> _device;
    MPSNNConvolution *convolution;
}

- (instancetype)initWithDevice:(id<MTLDevice>)device {
    self = [super init];
    if (self) {
        _device = device;
        // Setup Metal Performance Shader convolution kernel
        MPSNNConvolutionDescriptor *convDesc = [[MPSNNConvolutionDescriptor alloc] initWithKernelWidth:3
                                                                                          kernelHeight:3
                                                                                      inputFeatureChannels:1
                                                                                     outputFeatureChannels:1];
        convolution = [[MPSNNConvolution alloc] initWithDevice:_device
                                              convolutionDescriptor:convDesc];
    }
    return self;
}

- (void)performConvolutionWithInput:(MPSImage *)input
                             output:(MPSImage *)output {
    // Code to perform convolution
    // Example only: Ensure input/output handling is correct in actual code
    [convolution encodeToCommandBuffer:commandBuffer
                                sourceImage:input
                           destinationImage:output];
}

@end

File: templates/swift/cudnn_wrapper.swift
import MetalPerformanceShaders

class CUDNNWrapper {
    private let device: MTLDevice
    private var convolution: MPSCNNConvolution

    init(device: MTLDevice) {
        self.device = device

        let convDesc = MPSCNNConvolutionDescriptor(kernelWidth: 3, kernelHeight: 3,
                                                   inputFeatureChannels: 1, outputFeatureChannels: 1)

        convolution = MPSCNNConvolution(device: device, convolutionDescriptor: convDesc, kernelWeights: [], biasTerms: nil)
    }

    func performConvolution(input: MPSImage, output: MPSImage, commandBuffer: MTLCommandBuffer) {
        convolution.encode(commandBuffer: commandBuffer, sourceImage: input, destinationImage: output)
    }
}
